### PR TITLE
Add features for deploying static content

### PR DESCRIPTION
**Pull request by jmcphers**
This change adds support for deploying more content types from the rsconnect package. In particular:

1. The package is now able to deploy content to the RPubs service, via `rsconnect::rpubsUpload`. This function is a straightforward port of what was formerly an internal RStudio function, except that it now uses the HTTP transports provided by the rsconnect package, and writes deployment records which can be read by `rsconnect::deployments`. 
2. The package is now able to deploy static content to servers that support it. The static content mode is inferred when the set of files to be deployed includes no .Rmd files, no apparent Shiny layout, and at least one .html file. 
3. It is now possible to include custom metadata in the deployment record; the information is passed to `deployApp` and returned in `deployments`.

The new interface is backwards-compatible with the previous one.

### CODE REVIEW

File: R/bundle.R [hunk 1]
```diff
@@ -139,48 +104,63 @@ inferAppMode <- function(appDir, files) {
     return("rmd-static")
   }
 
-  # no Shiny .R files and no R Markdown docs--app mode is unknown
+  # if there are one or more HTML documents, use the static mode
+  htmlFiles <- grep("^[^/\\\\]+\\.html$", files, ignore.case = TRUE,
```
Comment (aronatkins) [inline R/bundle.R]: Do we ever see the `.htm` extension?

File: R/bundle.R [hunk 2]
```diff
@@ -188,27 +168,32 @@ createAppManifest <- function(appDir, appMode, accountInfo, files, appPrimaryRmd
...
+    # determine expected primary document extension
+    ext <- ifelse(appMode == "static", "html", "Rmd")
+
+    # use index file if it exists
+    primary <- which(grepl(paste0("^index\\.", ext, "$"), files, fixed = FALSE,
+                           ignore.case = TRUE))
     if (length(primary) == 0) {
       # no index.Rmd, so pick the first Rmd we find
-      primary <- which(grepl(glob2rx("*.Rmd"), files, fixed = FALSE, ignore.case = TRUE))
+      primary <- which(grepl(glob2rx("*.Rmd"), files, fixed = FALSE,
```
Comment (aronatkins) [inline R/bundle.R]: Should this search use the `ext` based on `appMode`?

File: R/bundle.R (hunk 1 above)
Comment (jjallaire) [inline R/bundle.R]: We certainly could (I can't vouch that the whole toolchain never produces
.htm and it seems likely that it does in at least one place). I'd
definitely support making the "l" optional here.

Comment (jmcphers) [inline R/bundle.R]: Good idea; I'll make this change.

File: R/bundle.R (hunk 2 above)
Comment (jmcphers) [inline R/bundle.R]: Yes, it should! Nice catch, fixed.

**Pull request merged**